src/goto-instrument/nondet_volatile.cpp

```diff
@@ -76,66 +99,76 @@ bool nondet_volatilet::is_volatile(const namespacet &ns, const typet &src)
   return false;
 }
 
+void nondet_volatilet::handle_volatile_expression(
+  exprt &expr,
+  const namespacet &ns)
+{
+  if(all_nondet ||
+    (expr.id() == ID_symbol &&
+    nondet_variables.count(to_symbol_expr(expr).get_identifier()) != 0))
+  {
+    typet t = expr.type();
+    t.remove(ID_C_volatile);
+
+    // replace by nondet
+    side_effect_expr_nondett nondet_expr(t, expr.source_location());
+    expr.swap(nondet_expr);
+  }
+}
+
 void nondet_volatilet::nondet_volatile_rhs(
   const symbol_tablet &symbol_table,
-  exprt &expr,
-  havoc_predicatet should_havoc)
+  exprt &expr)
 {
   Forall_operands(it, expr)
-    nondet_volatile_rhs(symbol_table, *it, should_havoc);
+    nondet_volatile_rhs(symbol_table, *it);
 
   if(expr.id()==ID_symbol ||
      expr.id()==ID_dereference)
   {
     const namespacet ns(symbol_table);
-    if(is_volatile(ns, expr.type()) && should_havoc(expr))
-    {
-      typet t=expr.type();
-      t.remove(ID_C_volatile);
 
-      // replace by nondet
-      side_effect_expr_nondett nondet_expr(t, expr.source_location());
-      expr.swap(nondet_expr);
+    if(is_volatile(ns, expr.type()))
+    {
+      handle_volatile_expression(expr, ns);
     }
   }
 }
 
 void nondet_volatilet::nondet_volatile_lhs(
   const symbol_tablet &symbol_table,
-  exprt &expr,
-  havoc_predicatet should_havoc)
+  exprt &expr)
 {
   if(expr.id()==ID_if)
   {
-    nondet_volatile_rhs(symbol_table, to_if_expr(expr).cond(), should_havoc);
+    nondet_volatile_rhs(symbol_table, to_if_expr(expr).cond());
```

Comment:
:question:  Is this case match already covered by existing tests somewhere?

Reply:
This is currently untested as the original nondet volatile code on which this builds was untested. I'll add more tests for the existing code in a separate PR.